nixos/modules/services/misc/ethminer.nix

```diff
@@ -0,0 +1,114 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.ethminer;
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    services.ethminer = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "enable ethminer ether mining.";
+      };
+
+      recheck = mkOption {
+        type = types.int;
+        default = 2000;
+        description = "time in milliseconds to farm recheck.";
+      };
+
+      toolkit = mkOption {
+        type = types.enum [ "cuda" "opencl" ];
+        default = "cuda";
+        description = "cuda or opencl toolkit.";
+      };
+
+      apiPort = mkOption {
+        type = types.int;
+        default = -3333;
+        description = "ethminer api port. minus sign puts api in read-only mode.";
+      };
+
+      wallet = mkOption {
+        type = types.string;
+        example = "0x0484A1EfB3C88B0C7dE21919D480fF51099c8113";
+        description = "ethereum wallet address.";
+      };
+
+      pool = mkOption {
+        type = types.string;
+        example = "eth-us-east1.nanopool.org";
+        description = "mining pool address.";
+      };
+
+      stratumPort = mkOption {
+        type = types.int;
+        default = 9999;
+        description = "stratum protocol tcp port.";
+      };
+
+      rig = mkOption {
+        type = types.string;
+        default = "mining-rig-name";
+        description = "mining rig name.";
+      };
+
+      notify = mkOption {
+        type = types.string;
+        example = "email%40example.org";
+        description = "url encoded email address to register with pool.";
+      };
+
+      power = mkOption {
+        type = types.int;
+        default = 115;
+        description = "miner max watt usage.";
+      };
+
+    };
+
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.ethminer = {
+      path = [ pkgs.cudatoolkit ];
+      description = "ethminer ethereum mining service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      serviceConfig = {
+        DynamicUser = true;
+        ExecStartPost = "+${config.boot.kernelPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.power}";
```

Comment:
What does this do and why is it needed? Also this probably only works for nvidia GPU's, so shouldn't this be conditioned on that?

And you should use `getBin config.boot.kernelPackages.nvidia_x11` instead.

Reply:
this throttles the power used by nvidia GPU's, as they do not need the default consumption to achieve full mining ability.  the limit does vary by nvidia model, and the default i set is for the 1070. This is only applicable to nvidia cards, so i will make that conditional.

Reply:
i made this conditional on cuda toolkit being specified `ExecStartPost = optional (cfg.toolkit == "cuda") "+${config.boot.kernelPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.maxPower}";`